### PR TITLE
Geometry test fix up

### PR DESCRIPTION
**Pull request by johntwyman**

### CODE REVIEW

File: CRM/CiviGeometry/BAO/Geometry.php [hunk 1]
```diff
@@ -265,8 +265,63 @@ public function calculateDistance($params) {
       1 => [$params['geometry_a'], 'String'],
       2 => [$params['geometry_b'], 'String'],
     ]);
-    $meters = (float) $result * 1000;
-    return $meters;
+    $metres = (float) $result * 1000;
+    return $metres;
+  }
+
```
Comment (andrew-cormick-dockery) [inline CRM/CiviGeometry/BAO/Geometry.php]: I think this needs a comment to say exactly what kind of spatial information you're returning.

Why is it all returning in a single function as opposed to using multiple functions, one for each piece of information?

File: CRM/CiviGeometry/BAO/Geometry.php [hunk 2]
```diff
@@ -265,8 +265,63 @@ public function calculateDistance($params) {
...
+  /**
+   * Output GeoJSON in a customFormat
+   * @param json $geoJSON - GeoJSON content to be passed out
+   */
+  public static function outputGeometryInFormat($geoJSON, $format) {
+    self::safeLoadAutoloader();
+    $polygon = geoPHP::load($geoJSON, 'json');
+    return $polygon->out($format);
+  }
+
```
Comment (andrew-cormick-dockery) [inline CRM/CiviGeometry/BAO/Geometry.php]: Haha... this was one of those things where I was wondering if we can get rid of it.  Why do we need envelopes?

File: api/v3/Geometry.php [hunk 3]
```diff
@@ -80,11 +98,24 @@ function civicrm_api3_geometry_delete($params) {
  * @throws API_Exception
  */
 function civicrm_api3_geometry_get($params) {
-  $results = _civicrm_api3_basic_get(_civicrm_api3_get_BAO(__FUNCTION__), $params);
+  if (!empty($params['format']) && !in_array($params['format'], ['json', 'kml', 'wkt'])) {
+    throw new API_Exception(E::ts('Output Format is not of an acceptable format it must be one of json, kml, or wkt'));
```
Comment (andrew-cormick-dockery) [inline api/v3/Geometry.php]: "Output format must be one of json, kml or wkt".  Too wordy.

File: api/v3/Geometry.php [hunk 4]
```diff
@@ -80,11 +98,24 @@ function civicrm_api3_geometry_delete($params) {
...
+  }
+  $sql = NULL;
+  if (!empty($params['geometry_collection_id'])) {
+    $geometries = civicrm_api3('Geometry', 'getcollection', ['collection_id' => $params['geometry_collection_id'], 'return' => ['geometry_id']]);
+    $geometryIds = CRM_Utils_Array::collect('geometry_id', $geometries['values']);
+    $sql = CRM_Utils_SQL_Select::fragment()->where('id IN (#geometryIDs)', ['geometryIDs' => $geometryIds]);
+  }
+  // Note we append additional SQL where clause here if geometry_collection_id is specfieid, this is a pseudo field
```
Comment (andrew-cormick-dockery) [inline api/v3/Geometry.php]: sp specified

File: tests/phpunit/api/v3/GeometryTest.php [hunk 5]
```diff
@@ -102,6 +103,21 @@ public function testCreateGeometry() {
     $geometryParams['geometry_type_id'] = "{$geometryType2['id']}, {$this->stateGeometryType['id']}";
     $this->callAPIFailure('Geometry', 'create', $geometryParams);
     $geometryParams['geometry_type_id'] = [$geometryType2['id'], $this->stateGeometryType['id']];
+    $sa1JSON = file_get_contents(\CRM_Utils_File::addTrailingSlash($this->jsonDirectoryStore) . '12101139836.json');
```
Comment (andrew-cormick-dockery) [inline tests/phpunit/api/v3/GeometryTest.php]: ooo, is this a file we've created for the purposes of testing?  can we re-name it something more meaningful than that?

File: tests/phpunit/api/v3/GeometryTest.php [hunk 6]
```diff
@@ -271,17 +292,20 @@ public function testRemoveOnlyCollection() {
...
    */
   public function testRemoveCollection() {
+    // Create Upper house Geometry
     $collectionParams = [
       'label' => 'Tasmanian Upper House',
       'source' => 'TasEC',
       'geometry_collection_type_id' => $this->externalCollectionType['id'],
     ];
     $collection = $this->callAPISuccess('GeometryCollection', 'create', $collectionParams);
+    // Create 2nd collection more broader than just about tasmanian upper house
```
Comment (andrew-cormick-dockery) [inline tests/phpunit/api/v3/GeometryTest.php]: How about "Create a superset of the first collection", which is what you're trying to do.  That it's Tassie is irrelevant.

File: tests/phpunit/api/v3/GeometryTest.php [hunk 7]
```diff
@@ -336,17 +365,20 @@ public function testGetGeometryCentroid() {
    * Test Archiving a Geometry
    */
   public function testArchiveGeometry() {
+    // Create Collection for wards
```
Comment (andrew-cormick-dockery) [inline tests/phpunit/api/v3/GeometryTest.php]: I think you should review these comments to make them more generic.  Certainly the tests are specific real-world examples of how the code might be used, but the test itself is more generic.  For example, this comment could be "Create collection to test the archiving function" could be used.  That it happens to be Qld Wards is neither here nor there - it could be anything.  You could even call it "Test Archive Collection".

File: tests/phpunit/api/v3/GeometryTest.php [hunk 8]
```diff
@@ -492,4 +535,68 @@ public function testGetDistance() {
     $this->assertEquals('2197', (int) $result['values']);
   }
 
+  /**
+   * Test returning spatial properties for a geometry
+   */
```
Comment (andrew-cormick-dockery) [inline tests/phpunit/api/v3/GeometryTest.php]: I'd like to try to avoid as much as possible tests which may change from time to time.  I don't know how often NSW branch boundaries change, but it certainly happens in other states.  Ideally we'd construct a fake geometry ourselves with known coordinates for the purpose of testing, but if we're going to take a real-life boundary then can we make sure the one we choose almost never change?  Like a state boundary for example.

Comment (seamuslee001) [reply tests/phpunit/api/v3/GeometryTest.php]: @andrew-cormick-dockery well we effectively are doing that we have the geometries that are used in theses tests in the load subfolder so in effect they are fake so to speak so these tests will live beyond any redistribution

File: tests/phpunit/api/v3/GeometryTest.php (hunk 5 above)
Comment (seamuslee001) [reply tests/phpunit/api/v3/GeometryTest.php]: Correct all the geometry that we are loading in is in the load sub folder of tests/phpunit/api/v3/

File: api/v3/Geometry.php (hunk 3 above)
Comment (seamuslee001) [reply api/v3/Geometry.php]: Do you have an alternate suggestion? The acceptable values of format is 'json', 'kml', 'wkt' perhaps?

File: CRM/CiviGeometry/BAO/Geometry.php (hunk 2 above)
Comment (seamuslee001) [reply CRM/CiviGeometry/BAO/Geometry.php]: These boundaries are used for zooming the various interactive maps i believe hence why i have sought to retain them here but we don't need to store the values in a table

Comment (andrew-cormick-dockery) [reply CRM/CiviGeometry/BAO/Geometry.php]: I guess that kind of denormalisation could be useful if there are some performance gains in terms of producing such a map for the end user.  At least, let's make this denormalisation automatic so that I don't have to do it myself when updating poly boundaries.

File: api/v3/Geometry.php (hunk 3 above)
Comment (andrew-cormick-dockery) [reply api/v3/Geometry.php]: My suggestion was as in my original comment "Output format must be one of json, kml or wkt"

File: tests/phpunit/api/v3/GeometryTest.php (hunk 5 above)
Comment (andrew-cormick-dockery) [reply tests/phpunit/api/v3/GeometryTest.php]: OK then let's rename the file so that it's clear what it is and what test it's used for.  The number alone provides no context as to its purpose.
